aten/src/ATen/native/LossMultiMargin.cpp

```diff
@@ -0,0 +1,328 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename scalar_t>
+inline scalar_t multi_margin_inner_sum_cpu(
+    const scalar_t* input_data,
+    const scalar_t* weight_data,
+    const int p,
+    const scalar_t margin,
+    const int64_t dim,
+    const int64_t target_idx) {
+  const scalar_t input_target = input_data[target_idx];
+  scalar_t sum = 0;
+  for (int64_t d = 0; d < dim; d++) {
+    if (d == target_idx) {
+      continue;
+    }
+
+    const scalar_t z = margin - input_target + input_data[d];
+    if (z > 0) {
+      scalar_t h = (p == 1) ? z : z * z;
+      if (weight_data != nullptr) {
+        h *= weight_data[target_idx];
+      }
+      sum += h;
+    }
+  }
+
+  sum /= dim;
+  return sum;
+}
+
+template <typename scalar_t>
+static inline void multi_margin_loss_cpu_kernel(
+    scalar_t* output_data,
+    scalar_t* input_data,
+    int64_t* target_data,
+    const int p,
+    scalar_t margin,
+    scalar_t* weight_data,
+    const int64_t nframe,
+    const int64_t dim,
+    const int64_t reduction) {
+  if (reduction == Reduction::None) {
+    for (int64_t t = 0; t < nframe; t++) {
+      auto sum = multi_margin_inner_sum_cpu(
+          input_data, weight_data, p, margin, dim, target_data[t]);
+      output_data[t] = sum;
+      input_data += dim;
+    }
+  } else {
+    scalar_t sum = 0;
+    for (int64_t t = 0; t < nframe; t++) {
+      sum += multi_margin_inner_sum_cpu(
+          input_data, weight_data, p, margin, dim, target_data[t]);
+      input_data += dim;
+    }
+    if (reduction == Reduction::Mean) {
+      sum /= nframe;
+    }
+    output_data[0] = sum;
+  }
+}
+
+void multi_margin_loss_out_cpu_template(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& target,
+    int p,
+    Scalar margin,
+    const Tensor& weight,
+    int64_t reduction) {
+  const auto ndims = input.dim();
+  TORCH_CHECK(
+      input.numel() > 0 && ndims <= 2,
+      "non-empty vector or matrix expected, got size: ",
+      input.sizes());
+
+  TORCH_CHECK(p == 1 || p == 2, "only p == 1 and p == 2 supported");
+
+  int64_t nframe, dim;
+  if (ndims <= 1) {
+    nframe = 1;
+    dim = ndims == 0 ? 1 : input.size(0);
```

Comment:
In a non-slavish port, we should not do this, and not special case the zero-dimensional case. But this is always ok to fix later.

Reply:
ok, you mean it would be better to drop support for scalar-tensors? I saw that the TH impl (accidentially) "supported" it and I did not want to introduce a breaking change.

Reply:
I think the docs are fairly clear that scalars aren't supposed to work. But yeah, I'd leave the breaking change for another PR.